### ISSUE TITLE
problem with derivative and chain rule

### ISSUE THREAD

**Issue by RiccardoRossi** [diofant/diofant#376]
Dear Developers,

i have been trying out Diofant on my computer (ubuntu 16.04) and i have a little issue with chain rule in derivative.

The issue is, perhaps not surprisingly, the same on sympy and diofant, still i thought it might be worth to post it here

from diofant import *

u,B = symbols('u B')
e = B*u #here i use an expression instead of a function

s = Function('s')(e)
print(diff(s,u).doit())

gives
B*Subs(Derivative(s(_xi_1), _xi_1), (_xi_1,), (B*u,))

while i would expect smthg like 
B*Derivative(s(e),e)

please be patient if the question/issue is wrong.

full question in the sympy forum is here:
https://groups.google.com/forum/#!msg/sympy/ogz52F5Ol5c/x2Mbwq41BAAJ

**Comment by skirpichev** [comment]
> while i would expect smthg like
> B*Derivative(s(e),e)

Actually, what I get here - is exactly that you expect.  Maybe with pretty-printing in usual mathematical notation it would be clear:
```
In [6]: diff(s,u).doit()
Out[6]: 
  ⎛ d        ⎞│      
B⋅⎜───(s(ξ₁))⎟│      
  ⎝dξ₁       ⎠│ξ₁=B⋅u
```

Please keep in mind, that in the diofant - there is no derivative operator, like Derivative in Mathematica.  You have to provide arguments, with respect to which you differentiate your expression.  That's why we have a Subs (unevaluated substitution) instance here and then - a Derivative instance inside, with a dummy variable.

> please be patient if the question/issue is wrong.

Don't worry.  Please take look on docstrings of classes Derivative, Subs; diff and doit methods.

Let me know if think that the answer is not clear enough.

> full question in the sympy forum is here:
> https://groups.google.com/forum/#!msg/sympy/ogz52F5Ol5c/x2Mbwq41BAAJ

It's not entirely clear for me what you are trying here to implement.

**Comment by RiccardoRossi** [comment]
Dear Sergey,
    thank you very much for your attention. I managed to reproduce the output u posted using jupyter-notebook, unfortunately it does not fit to what i am trying to achieve.

The thing is that i am developing a small tool to obtain the c++ implementation of a "Finite Element" starting from the symbolic definition of the underlyining functional.
From the point of view of symbolic calculus i think this is a rather simple thing since it only requires taking derivatives of polynomial functions.

My problem right now is that in some cases i don't want to do the symbolic derivation of some terms, but rather i would like to use a prescribed value of the derivative

in particular i have a term (known as internal forces, in my field) which assumes a form of the type

fint = B*s

where B is say a constant (but could also depend on a variable u) 

s := s(e(u))

where for example 
e = B*u #same B as before

I don't have available the explicit expression of the function s(e). This are actually calculated in an external library and i must take them for granted for modularity (it is someone else's work), however i can assume known (as symbols) the values of
s(e)
Derivative(s(e)),e) --> let's call this Ds_De, a constant whose value i can take for granted


note that for me it would be perfectly acceptable to be able to replace the "Subs(Derivative(s(_xi_1), _xi_1), (_xi_1,), (Bu,))" in the out expression by DsDe, just i am not able to match it ...



Trying following a different path, what i would like to achieve, in "pseudo code" is something like

u,B, DsDe = symbols('u B DsDe')
e = B*u 
s = FunctionWithDerivative('s')(e, DsDe)  
diff(s,u) --> should return  B*DsDe

Here "FunctionWithDerivative" would be an object such that 
diff(s,e) --> gives the symbol DsDe, which i assume known.

In the forum, i was suggested to implement my own class for FunctionWithDerivative, with a specific "fdiff" function, but i did something obviously wrong.

here is my code. It never enters the fdiff function...

class FunctionWithDerivative(Function):
    nargs = 3
   
    #first argument  --> the expression
    #second argument --> the variable it depends on
    #second argument --> the derivative      
    @classmethod
    def eval(cls,expression,x,D):
        print("--------- in eval -----------------. Looks like here it is not creating the object ...")
        cls.expression = expression
        cls.x = x
        cls.D = D
        return cls.expression

    def fdiff(self, argindex=1):
        print("********** inside fdiff ****",self.args)

       

u = symbols('u')
e = Function('e')(u)
s = FunctionWithDerivative(symbols('e'), symbols('e'), symbols('DsDe') )

print(diff(s,e).doit())

**Comment by skirpichev** [comment]
> It never enters the fdiff function...

Because your object was evaluated with eval method.  ``expression`` was returned instead of an instance of your class.  ``eval`` method - for Function evaluation, not for it's creation.

I still do not understand that the last expression expected to return, if not zero.  Your ``s`` and ``e`` - doesn't have common symbols.  But below is an example, where you can get unevaluated derivative and code enters ``fdiff`` method:
```
from diofant import *
from diofant.core.function import ArgumentIndexError


class FunctionWithDerivative(Function):
   nargs = 3

   @classmethod
   def eval(cls, expression, x, D):
       print("--------- in eval -----------------")
       cls.expression = expression
       cls.x = x
       cls.D = D

   def fdiff(self, argindex=1):
       print("********** inside fdiff ****", self.args)
       raise ArgumentIndexError


u = symbols('u')
e = Function('e')(u)
pprint(e)
s = FunctionWithDerivative(symbols('e'), symbols('e'), symbols('DsDe') )
pprint(s)
ds = diff(s, e).doit()
pprint(ds)  # <-- 0, as expected

ds2 = diff(s, symbols('e'))
pprint(ds2)
```

Output:
```
e(u)
--------- in eval -----------------
FunctionWithDerivative(e, e, DsDe)
0
********** inside fdiff **** (Symbol('e'), Symbol('e'), Symbol('DsDe'))
--------- in eval -----------------
--------- in eval -----------------
********** inside fdiff **** (Symbol('e'), Symbol('e'), Symbol('DsDe'))
--------- in eval -----------------
--------- in eval -----------------
⎛ ∂                                      ⎞│       ⎛ ∂                                      ⎞│    
⎜───(FunctionWithDerivative(e, ξ₂, DsDe))⎟│     + ⎜───(FunctionWithDerivative(ξ₁, e, DsDe))⎟│    
⎝∂ξ₂                                     ⎠│ξ₂=e   ⎝∂ξ₁                                     ⎠│ξ₁=e
```

More generic way of derivative evaluation customization - with ``_eval_derivative`` helper.  In this case, probably you would use ``Expr`` as a parent class.

> note that for me it would be perfectly acceptable to be able to replace the "Subs(Derivative(s(_xi_1), _xi_1), (_xi_1,), (Bu,))" in the out expression by DsDe, just i am not able to match it ...

Yes.  I think it's a bug.  It seems this expected to work with ``subs`` method if you replace ``Subs(Derivative(s(x), x), (x,), (B*u))`` (so, your matching expression could have any bounded symbol ``x``) with symbol ``DsDe``.

**Comment by skirpichev** [comment]
Issue with substitution: #377

**Comment by RiccardoRossi** [comment]
Hello again,

i have 95% of what i want working. I now miss to be able to do something like

diff(s,e), where e:=B*u (and B and u are symbols)

thing is that an error is thrown somewhere in function.py before my function is actually called. The error is apparently intentional, however in my case the derivation rule is so stuipid that i don't think it is possible to make it wrong:
derivative is either D (if we derive wrt the function variable) or 0 wrt any other var...

how can i avoid that error being thrown?

here is my code

from diofant import *

    class FunctionWithDerivative(Function):
            nargs = 2

            @classmethod
            def eval(cls, expression,  D):
                cls.expression = expression
                cls.D = D

            #added this function in an attempt to solve the last error
            def _eval_derivative(self, s):
                # f(x).diff(s) -> x.diff(s) * f.fdiff(1)(s)
                i = 0
                l = []
                for a in self.args:
                    i += 1
                    da = a.diff(s)
                    if da is S.Zero:
                        continue
                    try:
                        df = self.fdiff(i)
                    except ArgumentIndexError:
                        df = Function.fdiff(self, i)
                    l.append(df * da)
                return Add(*l)

            def fdiff(self, argindex=1):
                if(self.expression == self.args[0]):
                    return self.D
                else:
                    return self.D*diff(self.expression,self.args[0])



    #this works as i expected (maybe i should say, hoped)
    print("------------------ until here e(u) is a function -------------------------")
    u = symbols('u')
    e = Function('e')(u)
    s = FunctionWithDerivative(e, symbols('DsDe') )

    ds = diff(s, e).doit()
    pprint(ds)  # <-- DsDe, as expected
    print(diff(s,u).doit()) # <-- DsDe*Derivative(e(u), u), as expected

    print("------------------ starting from here e(u) = B*u -------------------------")


    ##now i redefine e to be an expression instead of a function
    B = symbols('B')
    e = B*u 
    s = FunctionWithDerivative(e, symbols('DsDe') ) #overwriting the old definition

    pprint(diff(s,u).doit()) # <-- B*DsDe as expected

    #NEXT should give back simply DsDe, however it does not enter in fdiff. an error is thrown before
    pprint(diff(s,e).doit())

**Comment by skirpichev** [comment]
> I now miss to be able to do something like diff(s,e), where e:=B*u (and B and u are symbols)

I'm afraid, but this is not going to be working.  You can't (in diofant, as well as in Mathematica) differentiate wrt general expressions; I believe - for good reasons.  For details you may read the Derivative docstring and sympy/sympy#460.

So, I doubt you can avoid this behavior without customized Derivative implementation.

PS: Probably, right now you can override stuff in ``Derivative.__new__`` this with ``_eval_diff`` helper for FunctionWithDerivative, but I'm not sure if we should keep this interface.  Anyway, ``diff`` function and ``Expr.diff`` method should be aliases, which is not the case.

**Comment by skirpichev** [comment]
Meanwhile, I'll close this issue.  Please reopen, if you think that have further questions on this topic.

Open a new issue (or pr request) if you think that something from discussed above - in the diofant is wrong (i.e. Derivative should behave differently or documentation could be improved).

**Comment by RiccardoRossi** [comment]
Dear Serge,

I read the link and the discussion makes complete sense to me.

I ll do without. I saw you committed a correction for the subs of the
derivative, I ll check it when I can.

Will your fix available through pip install?

Regards
Riccardo

**Comment by skirpichev** [comment]
> I saw you committed a correction for the subs of the derivative, I ll check it when I can.

This is not merged to master yet, but you can test fix in #378.

> Will your fix available through pip install?

If you asking for releasing minor versions like 0.8.1 - no, I'm not planning this soon.

But you can install development version from git (this [explained](http://diofant.readthedocs.io/en/stable/install.html#from-sources) in the documentation) or even with [pip install](https://pip.readthedocs.io/en/stable/reference/pip_install/#git), e.g. ``pip install git+https://github.com/diofant/diofant.git`` to install from the current master.  (You should have pip>=6.0.)

**Comment by skirpichev** [comment]
Now alpha release 0.9.0a2, including discussed here fix - is on pypi.  You could install one with ``--pre`` argument for ``pip install``.
